File: Lab 7/task.py
import time
from queue import Queue

class Process:
    def __init__(self, id, priority, burst_time, arrival_time):
        self.id = id
        self.priority = priority
        self.burst_time = burst_time
        self.arrival_time = arrival_time
        self.remaining_time = burst_time
        self.waiting_time = 0
        self.turnaround_time = 0

    def is_foreground(self):
        return self.priority >= 5

def calculate_waiting_time_priority(processes):
    processes.sort(key=lambda p: (-p.priority, p.arrival_time))
    time = 0
    for process in processes:
        if time < process.arrival_time:
            time = process.arrival_time
        process.waiting_time = time - process.arrival_time
        time += process.burst_time
        process.turnaround_time = process.waiting_time + process.burst_time

    avg_waiting_time = sum(p.waiting_time for p in processes) / len(processes)
    print("Average Waiting Time (Priority Scheduling):", avg_waiting_time)

def calculate_waiting_time_srtf(processes):
    processes.sort(key=lambda p: p.arrival_time)
    time = 0
    completed = 0
    ready_queue = []
    while completed < len(processes):
        # Add processes that have arrived
        for process in processes:
            if process.arrival_time <= time and process.remaining_time > 0 and process not in ready_queue:
                ready_queue.append(process)
        
        if ready_queue:
            ready_queue.sort(key=lambda p: p.remaining_time)
            current = ready_queue[0]
            time += 1
            current.remaining_time -= 1

            if current.remaining_time == 0:
                current.waiting_time = time - current.burst_time - current.arrival_time
                current.turnaround_time = current.waiting_time + current.burst_time
                completed += 1
                ready_queue.remove(current)
        else:
            time += 1

    avg_waiting_time = sum(p.waiting_time for p in processes) / len(processes)
    print("Average Waiting Time (SRTF):", avg_waiting_time)

def multi_level_feedback_queue(processes):
    foreground_queue = Queue()
    background_queue = Queue()

    for process in processes:
        if process.is_foreground():
            foreground_queue.put(process)
        else:
            background_queue.put(process)

    time = 0
    foreground = True
    while not foreground_queue.empty() or not background_queue.empty():
        if foreground and not foreground_queue.empty():
            print("Foreground Queue Execution - 10 seconds")
            for _ in range(10):
                if foreground_queue.empty():
                    break
                process = foreground_queue.get()
                print(f"Executing Process {process.id} with priority {process.priority}")
                time.sleep(1)  # Simulate 1 second of execution
                process.remaining_time -= 1
                if process.remaining_time > 0:
                    foreground_queue.put(process)  # Reinsert process with updated burst time
                else:
                    print(f"Process {process.id} completed")
                time += 1

        elif not foreground and not background_queue.empty():
            print("Background Queue Execution - 3 seconds")
            for _ in range(3):
                if background_queue.empty():
                    break
                process = background_queue.get()
                print(f"Executing Process {process.id} with remaining burst time {process.remaining_time}")
                time.sleep(1)  # Simulate 1 second of execution
                process.remaining_time -= 1
                if process.remaining_time > 0:
                    background_queue.put(process)  # Reinsert process with updated burst time
                else:
                    print(f"Process {process.id} completed")
                time += 1

        foreground = not foreground  # Alternate between foreground and background queues

if __name__ == "__main__":
    n = int(input("Enter the number of processes: "))
    processes = []

    for i in range(n):
        priority = int(input(f"Enter priority for process {i + 1}: "))
        burst_time = int(input(f"Enter burst time for process {i + 1}: "))
        arrival_time = int(input(f"Enter arrival time for process {i + 1}: "))
        processes.append(Process(i + 1, priority, burst_time, arrival_time))

    calculate_waiting_time_priority(processes)
    calculate_waiting_time_srtf(processes)
    multi_level_feedback_queue(processes)

File: Lab 7/task.cpp
#include <iostream>
#include <queue>
#include <vector>
#include <algorithm>
#include <unistd.h> // for sleep function
#include <climits>

using namespace std;

// Process structure to store process details
struct Process {
    int pid;        // Process ID
    int priority;   // Process priority
    int burstTime;  // Burst time
    int arrivalTime; // Arrival time
    int remainingTime; // Remaining burst time
};

class Scheduler {
private:
    vector<Process> processes;
    queue<Process> foregroundQueue;
    queue<Process> backgroundQueue;

public:
    // Method to input process details
    void inputProcessDetails() {
        int n;
        cout << "Enter number of processes: ";
        cin >> n;

        for (int i = 0; i < n; i++) {
            Process p;
            p.pid = i + 1;
            
            cout << "Process " << p.pid << " Details:" << endl;
            cout << "Priority (0-10): ";
            cin >> p.priority;
            
            cout << "Burst Time: ";
            cin >> p.burstTime;
            
            cout << "Arrival Time: ";
            cin >> p.arrivalTime;

            p.remainingTime = p.burstTime;
            processes.push_back(p);
        }
    }

    // Priority-based Scheduling
    void priorityScheduling() {
        // Sort processes by priority (higher priority first)
        sort(processes.begin(), processes.end(), 
            [](const Process& a, const Process& b) {
                return a.priority > b.priority;
            });

        int totalWaitingTime = 0;
        int currentTime = 0;

        for (auto& process : processes) {
            // Wait until process arrival time
            while (currentTime < process.arrivalTime) {
                currentTime++;
            }

            cout << "Executing Process " << process.pid 
                 << " (Priority: " << process.priority 
                 << ", Burst Time: " << process.burstTime << ")" << endl;

            // Calculate waiting time
            int waitingTime = currentTime - process.arrivalTime;
            totalWaitingTime += waitingTime;

            // Update current time
            currentTime += process.burstTime;
        }

        float avgWaitingTime = (float)totalWaitingTime / processes.size();
        cout << "Average Waiting Time (Priority Scheduling): " 
             << avgWaitingTime << endl;
    }

    // Shortest Remaining Time First (SRTF)
    void shortestRemainingTimeFirst() {
        int totalWaitingTime = 0;
        int currentTime = 0;
        int completedProcesses = 0;

        while (completedProcesses < processes.size()) {
            int shortestIndex = -1;
            int shortestTime = INT_MAX;

            // Find process with shortest remaining time
            for (int i = 0; i < processes.size(); i++) {
                if (processes[i].arrivalTime <= currentTime && 
                    processes[i].remainingTime > 0 && 
                    processes[i].remainingTime < shortestTime) {
                    shortestIndex = i;
                    shortestTime = processes[i].remainingTime;
                }
            }

            if (shortestIndex == -1) {
                currentTime++;
                continue;
            }

            // Execute the process
            processes[shortestIndex].remainingTime--;

            // Check if process is completed
            if (processes[shortestIndex].remainingTime == 0) {
                completedProcesses++;
                int completionTime = currentTime + 1;
                int turnaroundTime = completionTime - processes[shortestIndex].arrivalTime;
                int waitingTime = turnaroundTime - processes[shortestIndex].burstTime;
                totalWaitingTime += waitingTime;
            }

            currentTime++;
        }

        float avgWaitingTime = (float)totalWaitingTime / processes.size();
        cout << "Average Waiting Time (SRTF): " << avgWaitingTime << endl;
    }

    // Multilevel Feedback Queue
    void multiLevelFeedbackQueue() {
        // Separate processes into foreground and background queues
        for (auto& process : processes) {
            if (process.priority >= 5) {
                foregroundQueue.push(process);
            } else {
                backgroundQueue.push(process);
            }
        }

        int currentTime = 0;
        bool isForegroundTurn = true;
        int foregroundTimeSlice = 10;
        int backgroundTimeSlice = 3;

        while (!foregroundQueue.empty() || !backgroundQueue.empty()) {
            if (isForegroundTurn && !foregroundQueue.empty()) {
                // Execute foreground processes
                Process p = foregroundQueue.front();
                foregroundQueue.pop();

                int executionTime = min(p.remainingTime, foregroundTimeSlice);
                p.remainingTime -= executionTime;
                currentTime += executionTime;

                if (p.remainingTime > 0) {
                    foregroundQueue.push(p);
                }

                isForegroundTurn = false;
            } 
            else if (!backgroundQueue.empty()) {
                // Execute background processes
                Process p = backgroundQueue.front();
                backgroundQueue.pop();

                int executionTime = min(p.remainingTime, backgroundTimeSlice);
                p.remainingTime -= executionTime;
                currentTime += executionTime;

                if (p.remainingTime > 0) {
                    backgroundQueue.push(p);
                }

                isForegroundTurn = true;
            }
        }

        cout << "Multilevel Feedback Queue scheduling completed." << endl;
    }

    void runSchedulers() {
        cout << "\n--- Priority-based Scheduling ---" << endl;
        priorityScheduling();

        cout << "\n--- Shortest Remaining Time First ---" << endl;
        shortestRemainingTimeFirst();

        cout << "\n--- Multilevel Feedback Queue ---" << endl;
        multiLevelFeedbackQueue();
    }
};

int main() {
    Scheduler scheduler;
    scheduler.inputProcessDetails();
    scheduler.runSchedulers();

    return 0;
}